### PR TITLE
Refactor recent change to the thread pool to minimize binary size impact

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

A recent change (#21545) added a bool template parameter to a reasonably large class, causing a 16KB increase in the baseline minimal build.

Refactor to pass in the bool as a parameter to reduce the binary size growth to 1KB. 

### Motivation and Context

Minimal build binary size.

### CODE REVIEW

**Comment by skottmckay** [comment]
@goldsteinn could you please take a look in case there are any concerns.

File: include/onnxruntime/core/platform/threadpool.h [hunk 1]
```diff
@@ -424,8 +427,8 @@ class ThreadPool {
   ExtendedThreadPoolInterface* underlying_threadpool_ = nullptr;
 
   // If used, underlying_threadpool_ is instantiated and owned by the ThreadPool.
-  std::unique_ptr<ThreadPoolTempl<Env, true>> extended_eigen_hybrid_threadpool_;
-  std::unique_ptr<ThreadPoolTempl<Env, false>> extended_eigen_normal_threadpool_;
+  std::unique_ptr<HybridThreadPoolTempl<Env>> extended_eigen_hybrid_threadpool_;
+  std::unique_ptr<NormalThreadPoolTempl<Env>> extended_eigen_normal_threadpool_;
```
Comment (goldsteinn) [inline include/onnxruntime/core/platform/threadpool.h:431]: If you are dropping the template you no longer need two.

**Comment by goldsteinn** [comment]
So as you pointed on, the code size increase is from extra instantiated templates that could without a doubt be variables.
The reason I chose the template impl was to avoid any regressions.
We saw some models showings some regressions (although still net positive I'd say) with a more dynamic method.

Also note, the new code size will not have much effect on the footprint of used code size. We are instantiating two mutually exclusive worker impls, so while it wastes code size in the executable, only the one will ever be used (and contribute meaningfully to itlb/icache/etc..)

**Comment by snnn** [comment]
After the change #21545, we also see an increasing number of deadlocks happening in our pull request pipelines. Especially on ARM it is very common.  I am not sure if the two things are related.  It needs more investigation.

**Review by tianleiwu — APPROVED** [review]

**Comment by pranavsharma** [comment]
> After the change #21545, we also see an increasing number of deadlocks happening in our pull request pipelines. Especially on ARM it is very common. I am not sure if the two things are related. It needs more investigation. The deadlock issue existed before the change: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1485138&view=results

Did the number of deadlock occurrences _increase_ after #21545? If so, we need to revert that PR right away.

**Comment by pranavsharma** [comment]
> After the change #21545, we also see an increasing number of deadlocks happening in our pull request pipelines. Especially on ARM it is very common. I am not sure if the two things are related. It needs more investigation. The deadlock issue existed before the change: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1485138&view=results

Did the number of deadlock occurences _increase_ after #21545? This is one of the reasons I'm hesitant to touch any code near the threadpool class.

**Comment by skottmckay** [comment]
> The reason I chose the template impl was to avoid any regressions. We saw some models showings some regressions (although still net positive I'd say) with a more dynamic method.

Can you elaborate on how this could impact perf? 

If it's not templatized we have a switch statement in ThreadPoolWaiter but I expect that has a very low cost (one jump and very few instructions before the `return` when backoff is 0), especially in comparison to the call to SpinPause for the non-0 cases.

There are some ternary statements to choose the spin count and backoff which aren't constexpr anymore, however they're setting a const value using a const member so I would expect the compiler to be able to optimize those out (but low cost even if it can't).  We could make pref_spin_count/spin_count/steal_count const members if we thought the 3 ternary statements involved in setting those values had an impact on perf (which seems unlikely to me).

**Comment by snnn** [comment]
> Did the number of deadlock occurences increase after https://github.com/microsoft/onnxruntime/pull/21545?

Yes.  Now it happens every day. Previously I only saw it once in last month.

**Comment by pranavsharma** [comment]
> > Did the number of deadlock occurences increase after #21545?
> 
> Yes. Now it happens every day. Previously I only saw it once in last month.

We need to revert the PR right away; we should not let this go in the upcoming release.

**Comment by goldsteinn** [comment]
> > > Did the number of deadlock occurences increase after #21545?
> > 
> > 
> > Yes. Now it happens every day. Previously I only saw it once in last month.
> 
> We need to revert the PR right away; we should not let this go in the upcoming release.

Probably best while its debugged.
Is there any indication of what is causing the deadlock? The PR doesn't have any effect on the logic, just some tunings. If this is increasing deadlocks is likely means there is a deadlock condition in the `SetBlocked` codes that is independent from the PR. Lowering the iteration count just makes that code path more active.

**Comment by goldsteinn** [comment]
``

> > The reason I chose the template impl was to avoid any regressions. We saw some models showings some regressions (although still net positive I'd say) with a more dynamic method.
> 
> Can you elaborate on how this could impact perf?
> 
Well extra branches can obv impact perf. It was mostly noticable in first inference time.

**Comment by skottmckay** [comment]
> > Can you elaborate on how this could impact perf?
> 
> Well extra branches can obv impact perf. It was mostly noticable in first inference time.

There was noticeable slowdown in just the threadpool code or the inference as a whole was slower? 

The latter is expected as many things that are setup in the first inference are re-used for subsequent ones. e.g. memory pattern planning traces the first inference and enables using one block of memory for the next inference (we use offsets into that instead of individual allocation/free for the node outputs).

**Comment by tianleiwu** [comment]
> > Did the number of deadlock occurences increase after #21545?
> 
> Yes. Now it happens every day. Previously I only saw it once in last month.

I guess the cause pause_time_ was not initialized so it could wait a very long time. It is fixed in this PR.

**Comment by goldsteinn** [comment]
> > > Did the number of deadlock occurences increase after #21545?
> > 
> > 
> > Yes. Now it happens every day. Previously I only saw it once in last month.
> 
> I guess the cause: pause_time_ was not initialized so it could wait a very long time. It is fixed in this PR.

Is it always deadlocking or only sometimes? If its always then maybe (although constructing with `{}` default initializes it). If its sometimes then I don't think that would explain it.

**Comment by tianleiwu** [comment]
> > > > Did the number of deadlock occurences increase after #21545?
> > > 
> > > 
> > > Yes. Now it happens every day. Previously I only saw it once in last month.
> > 
> > 
> > I guess the cause: pause_time_ was not initialized so it could wait a very long time. It is fixed in this PR.
> 
> Is it always deadlocking or only sometimes? If its always then maybe (although constructing with `{}` default initializes it). If its sometimes then I don't think that would explain it.

I just reproduced one (normally, unit tests only took 1~2 minutes to finish):
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1514688&view=logs&j=ca395085-040a-526b-2ce8-bdc85f692774

Pipeline stops at the following unit test:
image

It is not related to initialization. Not sure whether it is related to dead lock.

**Comment by goldsteinn** [comment]
> > Did the number of deadlock occurences increase after #21545?
> 
> Yes. Now it happens every day. Previously I only saw it once in last month.

Are the deadlocks only happening on ARM? Or all platforms (just most common on ARM)?

**Comment by goldsteinn** [comment]
> > > > Did the number of deadlock occurences increase after #21545?
> > > 
> > > 
> > > Yes. Now it happens every day. Previously I only saw it once in last month.
> > 
> > 
> > I guess the cause: pause_time_ was not initialized so it could wait a very long time. It is fixed in this PR.
> 
> Is it always deadlocking or only sometimes? If its always then maybe (although constructing with `{}` default initializes it). If its sometimes then I don't think that would explain it.

Is there an outstanding issue regarding the monthly deadlock?
I tried reproducing it locally on a PI (because it was mentioned to be more common on ARM) and haven't hit
it in over 1000 runs of the test suite.

Are there any env / runtime factors that the deadlocking runners have which aren't default present for local builds?

**Pull request closed without merging**